pkg/protocol/xprotocol/dubbo/deconder.go

```diff
@@ -0,0 +1,129 @@
+package dubbo
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/AlexStocks/dubbogo/codec/hessian"
+	"mosn.io/mosn/pkg/buffer"
+	"mosn.io/mosn/pkg/types"
+)
+
+func decodeFrame(ctx context.Context, data types.IoBuffer) (cmd interface{}, err error) {
+	// convert data to duboo frame
+	dataBytes := data.Bytes()
+	frame := &Frame{}
+	// decode magic
+	frame.Magic = dataBytes[MagicIdx:FlagIdx]
+	// decode flag
+	frame.Flag = dataBytes[FlagIdx]
+	// decode status
+	frame.Status = dataBytes[StatusIdx]
+	// decode request id
+	reqIdRaw := dataBytes[IdIdx:(IdIdx + IdLen)]
+	frame.Id = binary.BigEndian.Uint64(reqIdRaw)
+	// decode data length
+	frame.DataLen = binary.BigEndian.Uint32(dataBytes[DataLenIdx:(DataLenIdx + DataLenSize)])
+
+	// decode event
+	eventBool := frame.Flag & (1 << 5)
+	if eventBool != 0 {
+		frame.Event = 1
+	} else {
+		frame.Event = 0
+	}
+	// decode twoway
+	twoWayBool := frame.Flag & (1 << 6)
+	if twoWayBool != 0 {
+		frame.TwoWay = 1
+	} else {
+		frame.TwoWay = 0
+	}
+	// decode direction
+	directionBool := frame.Flag & (1 << 7)
+	if directionBool != 0 {
+		frame.Direction = EventRequest
+	} else {
+		frame.Direction = EventResponse
+	}
+	// decode serializationId
+	frame.SerializationId = int(frame.Flag & 0x1f)
+
+	// decode payload
+	frame.payload = dataBytes[HeaderLen : HeaderLen+frame.DataLen]
```

Comment:
frame.rawData and frame.payload should point to an individual buffer, because dataBytes is actually the IO read buffer of current network connection. And might change after frame decoding, e.g. reuse IO read buffer would change the previous bytes.

Reply:
fixed